cdc/sink/ddlsink/mysql/async_ddl.go

```diff
@@ -0,0 +1,185 @@
+	tick := time.NewTimer(10 * time.Second)
+	defer tick.Stop()
+	log.Info("async exec add index ddl start",
+		zap.String("changefeedID", m.id.String()),
+		zap.Uint64("commitTs", ddl.CommitTs),
+		zap.String("ddl", ddl.Query))
+	go func() {
+		if err := m.execDDLWithMaxRetries(ctx, ddl); err != nil {
+			log.Error("async exec add index ddl failed",
+				zap.String("changefeedID", m.id.String()),
+				zap.Uint64("commitTs", ddl.CommitTs),
+				zap.String("ddl", ddl.Query))
+			done <- err
+			return
+		}
+		log.Info("async exec add index ddl done",
+			zap.String("changefeedID", m.id.String()),
+			zap.Uint64("commitTs", ddl.CommitTs),
+			zap.String("ddl", ddl.Query))
+		done <- nil
+	}()
+
+	select {
+	case <-ctx.Done():
+		// if the ddl is canceled, we just return nil, if the ddl is not received by tidb,
+		// the downstream ddl is lost, because the checkpoint ts is forwarded.
+		log.Info("async add index ddl exits as canceled",
+			zap.String("changefeedID", m.id.String()),
+			zap.Uint64("commitTs", ddl.CommitTs),
+			zap.String("ddl", ddl.Query))
+		return nil
+	case err := <-done:
+		// if the ddl is executed within 2 seconds, we just return the result to the caller.
+		return err
+	case <-tick.C:
+		// if the ddl is still running, we just return nil,
+		// then if the ddl is failed, the downstream ddl is lost.
+		// because the checkpoint ts is forwarded.
+		log.Info("async add index ddl is still running",
+			zap.String("changefeedID", m.id.String()),
+			zap.Uint64("commitTs", ddl.CommitTs),
+			zap.String("ddl", ddl.Query))
+		m.runningAsyncDDLs.Store(ddl.TableInfo.TableName, struct{}{})
+		return nil
+	}
+}
+
+// Should always wait for async ddl done before executing the next ddl.
+func (m *DDLSink) waitAsynExecDone(ctx context.Context, ddl *model.DDLEvent) {
+	if !m.cfg.IsTiDB {
+		return
+	}
+
+	tables := make(map[model.TableName]struct{})
+	if ddl.TableInfo != nil {
+		tables[ddl.TableInfo.TableName] = struct{}{}
+	}
+	if ddl.PreTableInfo != nil {
+		tables[ddl.PreTableInfo.TableName] = struct{}{}
+	}
+	if len(tables) == 0 || m.checkAsyncExecDDLDone(ctx, tables) {
+		return
+	}
+
+	log.Debug("wait async exec ddl done",
+		zap.String("namespace", m.id.Namespace),
+		zap.String("changefeed", m.id.ID),
+		zap.Any("tables", tables),
+		zap.Uint64("commitTs", ddl.CommitTs),
+		zap.String("ddl", ddl.Query))
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			done := m.checkAsyncExecDDLDone(ctx, tables)
+			if done {
+				return
+			}
+		}
+	}
+}
+
+func (m *DDLSink) checkAsyncExecDDLDone(ctx context.Context, tables map[model.TableName]struct{}) bool {
+	for table := range tables {
+		done := m.doCheck(ctx, table)
+		if !done {
+			return false
+		}
+	}
+	return true
+}
+
+func (m *DDLSink) doCheck(ctx context.Context, table model.TableName) (done bool) {
+	if _, ok := m.runningAsyncDDLs.Load(table); !ok {
```

Comment:
If there is a running async DDL after the CDC restarts, this check will fail to detect it here. 
A mark needs to be added to determine if the CDC has just restarted. The justRestarted state can transition to normal after one DDL is executed.

Reply:
we can not determine the state by single mark, suppose there are two tables t1 and t2, then:
1.  Before restart: t1->running ddl1
2.  After restart: if there is a ddl event of t2 executed before t1, then the justRestarted is useless.

Reply:
The purpose of adding caching here is to reduce the frequency of database reads. A fixed-size LRU seems like a reasonable choice.